File: companies/Google/130.surrounded-regions.go
package leetcode

// Solve130 marks all border-connected 'O' cells as safe with BFS, flips the
// remaining enclosed 'O' cells to 'X', then restores safe cells. Border-connected
// regions cannot be surrounded.
//
// Time: O(m*n). Space: O(m*n) worst-case queue.
func Solve130(board [][]byte) {
	if len(board) == 0 || len(board[0]) == 0 {
		return
	}
	rows, cols := len(board), len(board[0])
	type cell struct{ r, c int }
	q := []cell{}
	mark := func(r, c int) {
		if board[r][c] == 'O' {
			board[r][c] = 'S'
			q = append(q, cell{r, c})
		}
	}
	for r := 0; r < rows; r++ {
		mark(r, 0)
		mark(r, cols-1)
	}
	for c := 0; c < cols; c++ {
		mark(0, c)
		mark(rows-1, c)
	}
	dirs := [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	for head := 0; head < len(q); head++ {
		cur := q[head]
		for _, d := range dirs {
			nr, nc := cur.r+d[0], cur.c+d[1]
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols && board[nr][nc] == 'O' {
				board[nr][nc] = 'S'
				q = append(q, cell{nr, nc})
			}
		}
	}
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if board[r][c] == 'O' {
				board[r][c] = 'X'
			} else if board[r][c] == 'S' {
				board[r][c] = 'O'
			}
		}
	}
}

File: companies/Zomato/130.surrounded-regions.py
"""
Approach: Mark border-connected 'O' cells as safe, then flip every remaining 'O'.
Data structure: a deque performs iterative BFS and avoids recursion-depth issues on large boards.
Interview logic: an 'O' is uncapturable exactly when it can reach the border through other 'O' cells. Finding those safe cells first lets the rest be flipped confidently.
Complexity: O(mn) time, O(mn) worst-case queue space.
Tests and edge cases: empty board; all-border regions stay 'O'; fully enclosed regions become 'X'.
"""
from __future__ import annotations
from collections import deque
from typing import List

# @lc code=start
from collections import deque
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        if not board or not board[0]:
            return
        rows, cols = len(board), len(board[0])
        q = deque()
        def mark(r: int, c: int) -> None:
            if board[r][c] == 'O':
                board[r][c] = 'S'
                q.append((r, c))
        for r in range(rows):
            mark(r, 0); mark(r, cols - 1)
        for c in range(cols):
            mark(0, c); mark(rows - 1, c)
        while q:
            r, c = q.popleft()
            for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols and board[nr][nc] == 'O':
                    board[nr][nc] = 'S'
                    q.append((nr, nc))
        for r in range(rows):
            for c in range(cols):
                if board[r][c] == 'O':
                    board[r][c] = 'X'
                elif board[r][c] == 'S':
                    board[r][c] = 'O'
# @lc code=end
